# 📚 [MS Learn C#](https://learn.microsoft.com/en-us/users/dotnet/collections/yz26f8y64n7k07) 📚 

_______________________ 
## ***[Write your first code using C# <br> (Get started with C#, Part 1)](https://learn.microsoft.com/en-us/training/paths/get-started-c-sharp-part-1/)***

_________________________

### [Write your first C# code](https://learn.microsoft.com/en-au/training/modules/csharp-write-first/)

_________________________

####  ✔️ **Learning Objectives:**


* Write your first lines of C# Code
* Use two different techniques to print a message as output
* Diagnose errors when code is incorrect
* Identify different C# syntax elements like operators, classes and methods


#### **Write() & WriteLine()**

In [2]:
Console.WriteLine("My name is Kristy");
// At the end of the printed text, it adds a line. 
// Any new output will print on the new line below the previous output.

My name is Kristy


In [3]:
Console.Write("My");
Console.Write(" ");
Console.Write("name");
Console.Write(' ');
Console.Write('i');
Console.Write("s Kristy");
// Any new output will print on the same line of the previous output.

My name is Kristy

In [5]:
console.WriteLine("My name is Kristy");
// classes and methods are case sensitive

Error: (1,1): error CS0103: The name 'console' does not exist in the current context

In [15]:
Console.writeLine("My name is Kristy");

Error: (1,9): error CS0117: 'Console' does not contain a definition for 'writeLine'

In [16]:
Console.WriteLine("My Name is Kristy");
// Double Quotation = Literal String

My Name is Kristy


In [17]:
Console.WriteLine('My Name is Kristy');
// One character only in Character Literal

Error: (1,19): error CS1012: Too many characters in character literal

In [18]:
Console.WriteLine('M');
// character literal

M


#### **Programming Language & Compilation**

##### What is a Programming Language?
* Lets you write instructions you want computer to carry out
* allow humans to express intent in understandable way (computers can't understand code)
* code must be **compiled** into a formate the computer can understand


##### What is Compilation?
* Compiler converts source code into a format the CPU can execute
* CPU understands instructions by turning switches on and off (0 + 1)

#### **Class, Method & Member Access / Invocation / Statement Operators**

* `Console.WriteLine("Double Quotation");`
* Class.Method(input parameter);
* `.` --> member access operator
* `()` --> invocation operator
* `;` --> statement operator (tells compiler the command is done)


#### ❓ **Check Knowledge: Write() & WriteLine()**



* What is the difference between `Console.Write` and `Console.WriteLine`?
    * ❌ `Console.Write` prints the output on a new line. --> Prints output on the same line
    * ❌ `Console.WriteLine` prints the output on a new line. --> Prints output on the same line
    * ✔️ `Console.WriteLine` appends a new line after the output.

#### ⭐ **Challenge: Write code in the .NET Editor to display two messages**

In [10]:
Console.WriteLine("This is the first line");
Console.Write("This ");
Console.Write("is ");
Console.Write("the ");
Console.Write("second ");
Console.Write("line.");

This is the first line
This is the second line.

#### ❓ **Check Knowledge: Compiler, Literal String, Errors**


* What is the primary job of the compiler?
    * ❌ Locates spelling mistakes in your code
    * ❌ Executs your code
    * ✔️ Converts your code into an executable format that the computer can understand
    
    <br>

* Which of the following statements is true about C#?
    * ❌ Case insensitive --> Case Sensitive
    * ❌`Console`is a method, `WriteLine()` is a class --> `Console`=Class, `WriteLine()`=Method.
    * ✔️ Converts your code into an executable format that the computer can understand
    
    <br>

* What is wrong w/ this code: `Console.Writeine("What is wrong with this code?")`
    * ❌ The L in `WriteLine` should be lower-case --> uppercase
    * ✔️ It's missing a `;` (semi-colon) at the end
    * ❌ The string should use `''` (single quotes) --> double quotes

#### **SUMMARY:**

* Your goal was to write code that displayed simple messages to an output console while familiarizing yourself with the syntax. 
* You wrote your first lines of code using basic C# syntax. 
* You learned two techniques for displaying literal-string data to the console. 
* You also learned what to look for when you come across an error in your code. 
* And lastly, you identified C# syntax elements like classes and methods, and the purpose of several special symbols that are known as operators. 
* You've taken your first steps towards building more sophisticated applications.

_________________________

### [Store and retrieve data using literal and variable values in C#](https://learn.microsoft.com/en-us/training/modules/csharp-literals-variables/)

_________________________

####  ✔️ **Learning Objectives:**

* Create literal values for five basic data types
* Declare and initialize variables
* Retrieve and set values in variables
* Allow the compiler to determine the data type for your variable when initializing

####  **Literal Values: Str, Char, Int, Float, Bool**

* constant value that never changes
* `string`, `char`, `integer`, `boolean`, `float` (float, double, decimal)

In [24]:
// string literal
Console.WriteLine("string literal");

// character literal
Console.WriteLine('c');

string literal
c


In [37]:
// character literal w/ error:
Console.WriteLine('error');

Error: (2,19): error CS1012: Too many characters in character literal

In [40]:
// boolean literal
Console.WriteLine(true);
Console.WriteLine(false);

// vs String
Console.WriteLine("true");

True
False
true


In [39]:
// integer literal
Console.WriteLine(123);

// vs String
Console.WriteLine("123");

123
123


*Floating Point Literals*


| Float Type | Precision | Notes |
|-|-|-|
| float | ~6-9 digits | Fractional Values with `f` suffix |
| double  | ~15-17 digits| decimal number, compilaer defaults to double without suffix |
| decimal | 28-29 digits| tells compiler we are working w/ deximal type with `m` suffix |

In [31]:
// FLOAT W/ F
Console.WriteLine(0.123456789F);
Console.WriteLine(012.3456789F);

0.12345679
12.345679


In [32]:
// DOUBLE
Console.WriteLine(0.1234567890123456789);
Console.WriteLine(012.34567890123456789);

0.12345678901234568
12.345678901234567


In [36]:
// DECIMAL W/ m
Console.WriteLine(0.123456789012345678901234567890123m);
Console.WriteLine(012.3456789012345678901234567890123m);
d

0.1234567890123456789012345679
12.345678901234567890123456789


####  **Declare Variables (vs literal):**

|Literal|Declare Variable|
|-|-|
|Literally a hard-coded value | Not Hard-Coded, Temporary stored in computer memory |
|Constant / Unchanged while executing | Declare before use. Select data type and give variable name |

Flow:
* Declare -> set -> retrieve -> intialise

Declare a variable with data type and name of variable in camelCase


In [44]:
// Conventional:

string firstName; // camelCase (not PascalCase or snake_case)
string _firstName; // _ for special purpose only

char userOption;
int gameScore;
decimal particlesPerMillion;
bool processedCustomer;

In [43]:
// Non-Conventional (avoid)

string FirstName; // Pascal (use for Class naming only)
string #name;
string $name;
string string;
string 1name;
string notMeaningfulName;
string strValue;

Error: (5,1): error CS1519: Invalid token 'string' in class, record, struct, or interface member declaration
(4,8): error CS1040: Preprocessor directives must appear as the first non-whitespace character on a line
(5,8): error CS1519: Invalid token '$' in class, record, struct, or interface member declaration
(5,8): error CS1056: Unexpected character '$'
(6,8): error CS1519: Invalid token 'string' in class, record, struct, or interface member declaration
(6,14): error CS1519: Invalid token ';' in class, record, struct, or interface member declaration
(7,8): error CS1519: Invalid token '1' in class, record, struct, or interface member declaration
(7,9): error CS1002: ; expected

#### **Set (Assign), Get (Retrieve), Modify, Initialise Variables**

* variables = temp storage containers for data
* You must assign (set) a value to a variable before you can retrieve (get) a value from a variable.
* You can initialize a variable by assigning a value to the variable at the point of declaration.
* Assignment happens from right to left.
* You use a single equals character as the assignment operator.
* To retrieve the value from the variable, you merely use the variable's name.

In [50]:
// DECLARE, SET, PRINT
string firstName; // declare
firstName = "Bob"; // set
Console.WriteLine(firstName); // print

// " = " assignment operator

Bob


In [51]:
// INCORRECT DATA TYPE
int firstName;
firstName = "Bob";
Console.WriteLine(firstName); 

Error: (3,13): error CS0029: Cannot implicitly convert type 'string' to 'int'

In [53]:
// MODIFY
firstName = "New Name";
Console.WriteLine(firstName);

New Name


In [54]:
// INITIALISE
string newName = "initilised name";
Console.WriteLine(newName);

initilised name


#### **Declare Implicitly Typed Local Variables**

* `var` keyword + intialisation = implicitly typed local variable
* implies data type without explicitly typing it
* `var` used as time saver when type is obvious from context
* once value type is set, it's locked
* MUST initialise with value (or error is thrown)

In [67]:
var message;

Error: (1,5): error CS0818: Implicitly-typed variables must be initialized

In [55]:
var message = "Implicitly types local variable, string, initialised";
Console.WriteLine(message); // string type implied due to value

Implicitly types local variable, string, initialised


In [63]:
message = "value is mutable, data type is immutable";

In [64]:
message = 4;

Error: (1,11): error CS0029: Cannot implicitly convert type 'int' to 'string'

In [65]:
var newNumber = 4;
Console.WriteLine(newNumber);

4


In [66]:
newNumber = "4";

Error: (1,13): error CS0029: Cannot implicitly convert type 'string' to 'int'

#### ⭐ **Challenge: Display literal and variable values**

**[CLICK FOR MSLEARN SOLUTION](https://learn.microsoft.com/en-us/training/modules/csharp-literals-variables/7-solution)**

![Alt text](\Assets\image-2.png)

Output required:

`Hello, Bob! You have 3 messages in your inbox. The temperature is 34.4 celsius.`

Plan:

`Hello, {firstName}! You have {numberOfMessages} messages in your inbox. The temperature is {temperature} celsius.`


In [70]:
// Literal
string firstName;
firstName = "Bob";
Console.WriteLine(firstName);

int numberOfMessages = 3;
Console.WriteLine(numberOfMessages);

// Implicit
var temperature = 34.4m;
Console.WriteLine(temperature);

// Combine
Console.Write("Hello, ");
Console.Write(firstName);
Console.Write("! You have ");
Console.Write(numberOfMessages);
Console.Write(" messages in your inbox. The temperature is ");
Console.Write(temperature);
Console.Write(" celsius. ");


Bob
3
34.4
Hello, Bob! You have 3 messages in your inbox. The temperature is 34.4 celsius. 

#### ❓ **Check Knowledge: variable declaration, assignment, var, data type**


1. Which of the following variable names should or must be avoided? 
    * ✔️ $DATA
    * ❌ registrationComplete
    * ❌ flag

    <br>

2. What is the problem with this line of code: `var message`; 
    * ❌ var isn't a data type.
    * ✔️ Use of var keyword without initializing the variable
    * ❌ A variable named message should always be a string.

    <br>

3. Which of the following is the output of Console.WriteLine(34.40M);? 
    * ❌ 34.40M
    * ❌ 34.4
    * ✔️ 34.40

    <br>

4. Which of the following lines of code creates a variable correctly? 
    * ❌ int x = 12.3m;
    * ✔️ decimal x = 12.3m;
    * ❌ bool x = 'False';

#### **SUMMARY:**

* Your goal was to display a formatted message using a combination of literal and variable values.
* Using basic C# syntax, you created literal values of several different data types. 
* You declared variables, and also set and retrieved values from those variables. 
* You also initialized variables, and learned how to use the var keyword to implicitly type a variable by inferring the type from the initialization.

_________________________

### [Perform basic string formatting in C#](https://learn.microsoft.com/en-us/training/modules/csharp-basic-formatting/)

_________________________

####  ✔️ **Learning Objectives:**

* Create string data containing tabs, new lines, and other special characters
* Create string data containing Unicode characters
* Combine string data into a new string value via concatenation
* Combine string data into a new string value via interpolation*

####  **Character escape sequences - `\n , \u, \t, \\", \\\`***

* `\`   escape next character
* `\n`  new line
* `\t`  tab
* `\"`  escape double quotation mark
* `\\`  escape backslash
* `@`   create verbatim string literal to keep all whitespace and \ in string
* `\u`  + four character code for Unicode characters (UTF-16) in a string

In [1]:
Console.WriteLine("Hello\nNew Line!");
Console.WriteLine("Hello\tTab!");
Console.WriteLine("Hello \"Quotations\"!");
Console.WriteLine("c\\some\\path");

Hello
New Line!
Hello	Tab!
Hello "Quotations"!
c\some\path


In [2]:
// errors:
Console.WriteLine("Hello "Quotations"!");

Error: (2,27): error CS1003: Syntax error, ',' expected
(2,37): error CS1003: Syntax error, ',' expected

In [3]:
// error unrecognised escape sequence
Console.WriteLine("c:\some\path");

Error: (2,22): error CS1009: Unrecognized escape sequence
(2,27): error CS1009: Unrecognized escape sequence

In [4]:
Console.WriteLine("Generating invoices for customer \"Contoso Corp\" ...\n");
Console.WriteLine("Invoice: 1021\t\tComplete!");
Console.WriteLine("Invoice: 1022\t\tComplete!");
Console.WriteLine("\nOutput Directory:\t");

Generating invoices for customer "Contoso Corp" ...

Invoice: 1021		Complete!
Invoice: 1022		Complete!

Output Directory:	


#### **Verbatim string literal - `@`**

In [5]:
// @ - verbatim string literal
// keep all whitespace and characters without " \ "

Console.WriteLine(@"    c:\source\repos    
        (this is where your code goes)");



    c:\source\repos    
        (this is where your code goes)


In [7]:
Console.WriteLine("Generating invoices for customer \"Contoso Corp\" ...\n");
Console.WriteLine("Invoice: 1021\t\tComplete!");
Console.WriteLine("Invoice: 1022\t\tComplete!");
Console.WriteLine("\nOutput Directory:\t");
Console.Write(@"c:\invoices"); // @ allows use of " \ " without issue

Generating invoices for customer "Contoso Corp" ...

Invoice: 1021		Complete!
Invoice: 1022		Complete!

Output Directory:	
c:\invoices

#### **Unicode - `\u`**

In [ ]:
// \u = unicode escape characters (UTF-16)
// Kon'nichiwa World
Console.WriteLine("\u3053\u3093\u306B\u3061\u306F World!");

こんにちは World!


In [9]:
Console.WriteLine("Generating invoices for customer \"Contoso Corp\" ...\n");
Console.WriteLine("Invoice: 1021\t\tComplete!");
Console.WriteLine("Invoice: 1022\t\tComplete!");
Console.WriteLine("\nOutput Directory:\t");
Console.Write(@"c:\invoices");

// To generate Japanese invoices:
// Nihon no seikyū-sho o seisei suru ni wa:
Console.Write("\n\n\u65e5\u672c\u306e\u8acb\u6c42\u66f8\u3092\u751f\u6210\u3059\u308b\u306b\u306f\uff1a\n\t");
// User command to run an application
Console.WriteLine(@"c:\invoices\app.exe -j");

Generating invoices for customer "Contoso Corp" ...

Invoice: 1021		Complete!
Invoice: 1022		Complete!

Output Directory:	
c:\invoices

日本の請求書を生成するには：
	c:\invoices\app.exe -j


#### **String concatenation - `+`**

* allows you to combine smaller literal and variable strings into a single string
* avoid creating intermediate variables if it doesn't improve readability

In [10]:
string firstName = "Bob";
string message = "Hello " + firstName;
Console.WriteLine(message);

Hello Bob


In [13]:
string firstName = "Bob";
string greeting = "Hello";
string message = greeting + " " + firstName + "!"; // intermediate variable
Console.WriteLine(message);

Hello Bob!


In [12]:
// acoid intermediate variables
string firstName = "Bob";
string greeting = "Hello";
Console.WriteLine(greeting + " " + firstName + "!");

Hello Bob!


#### **String interpolation - `{variable}`**

* string interpolation can be better than concatenation - can reduce number of characters required
* combine interpolation (`{}`) and verbatim literals (`@`) in a string template
    * eg: `$"{variable}"`

In [16]:
// concatination
string message = greeting + " " + firstName + " -> concatination";
Console.WriteLine(message);

// interpolation
string message2 = $"{greeting} {firstName} -> interpolation";
Console.WriteLine(message2);

// combine literal string and variable value
string message3 = $"Hello {firstName} -> literal string + variable value";
Console.WriteLine(message3);

// combo: interpolation, multi-variables, literal strings
int version = 11;
string updateText = "Update to Windows";
string message4 = $"{updateText} {version}";
Console.WriteLine(message4 + "-> using intermeidate variable");

// avoid intermediate variable
// don't implement message4
Console.WriteLine($"{updateText} {version} -> avoiding intermediate variable");

// combo: verbatim literal, string interpolation
string projectName = "First-Project";
Console.WriteLine($@"C:\Output\{projectName}\Data");

Hello Bob -> concatination
Hello Bob -> interpolation
Hello Bob -> literal string + variable value
Update to Windows 11-> using intermeidate variable
Update to Windows 11 -> avoiding intermediate variable
C:\Output\First-Project\Data


#### ⭐ **Challenge: Format and Display Instructions**

**[CLICK FOR MSLEARN SOLUTION](https://learn.microsoft.com/en-us/training/modules/csharp-basic-formatting/6-solution)**

![Alt-text](./Assets/image-challenge-format-and-display-instructions.png)

In [29]:
// Initialise variables
// \u unicode

string projectName = "ACME";
string russianMessage = "\u041f\u043e\u0441\u043c\u043e\u0442\u0440\u0435\u0442\u044c \u0440\u0443\u0441\u0441\u043a\u0438\u0439 \u0432\u044b\u0432\u043e\u0434";

Console.WriteLine($"View English Output:\n\tc:\\Exercise\\{projectName}\\data.txt\n");

Console.WriteLine($"{russianMessage}:\n\t c:\\Exercise\\{projectName}\\ru-RU\\data.txt");

View English Output:
	c:\Exercise\ACME\data.txt

Посмотреть русский вывод:
	 c:\Exercise\ACME\ru-RU\data.txt


In [30]:
// MS Learn solution

string projectName = "ACME";
string englishLocation = $@"c:\Exercise\{projectName}\data.txt";
Console.WriteLine($"View English output:\n\t\t{englishLocation}\n");

string russianMessage = "\u041f\u043e\u0441\u043c\u043e\u0442\u0440\u0435\u0442\u044c \u0440\u0443\u0441\u0441\u043a\u0438\u0439 \u0432\u044b\u0432\u043e\u0434";
string russianLocation = $@"c:\Exercise\{projectName}\ru-RU\data.txt";
Console.WriteLine($"{russianMessage}:\n\t\t{russianLocation}\n");

View English output:
		c:\Exercise\ACME\data.txt

Посмотреть русский вывод:
		c:\Exercise\ACME\ru-RU\data.txt



#### ❓ **Check Knowledge: escape sequences, string interpolation, unicode**


1. Which of the following escape sequences should be used to add double quotes to a literal string in C# code?  
    * ❌ `\q`
    * ❌ `\`
    * ✔️ `\"`

    <br>

2. Which of the following lines of code correctly uses string interpolation assuming that the variable value is a string? 
    * ❌ `Console.WriteLine(@"My value: {value}");`
    * ✔️ `Console.WriteLine($"My value: {value}");`
    * ❌ `Console.WriteLine(@"My value: [value]");`

    <br>

3. Unicode characters were embedded in the C# strings for a console application to present a greeting message in Thai. However, the message is only displayed as question mark characters. What is a possible cause? 
    * ❌ A missing \u escape sequence for Unicode characters. --> only numbers will appear in this case
    * ❌ The wrong Unicode characters were used.
    * ✔️ The user's console doesn't support Unicode characters.

    <br>

4. Which of the following lines of code append a single \ to the string directory? 
    * ❌ `directory = directory + "\";`
    * ❌ `directory = directory + '\';`
    * ✔️ `directory = directory + @"\";`

#### **SUMMARY:**

* Your goal was to write code that formats strings with special characters, such as double quotes, new lines, tabs, and other white space, as well as unicode characters. 
* You also combined strings using two different techniques.
* Using character escape sequences, you added special characters in your literal strings by either using special escaping sequences or using verbatim strings. 
* You added Unicode characters from other language sets like Japanese Kanji and Russian Cyrillic in your literal strings. 
* You used simple string concatenation with the + symbol, and upgraded to string interpolation for combining values into a string template.
* Without the ability to format your output, you would be severely restricted in what kinds of information you could present to the user. 
* However, now you can provide your users with sophisticated instructions and feedback with a wide variety of formatting, symbols, and languages.

_________________________

### [Perform basic operations on numbers in C#](https://learn.microsoft.com/en-us/training/modules/csharp-basic-operations/)

_________________________

_________________________

### [Guided project - Calculate and print student grades](https://learn.microsoft.com/en-gb/training/modules/guided-project-calculate-print-student-grades/)

_________________________

_________________________

### [Guided project - Calculate final GPA](https://learn.microsoft.com/en-gb/training/modules/guided-project-calculate-final-gpa/)

_________________________

_________________________

## ***[Create and run simple C# console applications <br> (Get started with C#, Part 2)](https://learn.microsoft.com/en-us/training/paths/get-started-c-sharp-part-2/)***

_________________________

_________________________

## ***[Add logic to C# console applications <br> (Get started with C#, Part 3)](https://learn.microsoft.com/en-us/training/paths/get-started-c-sharp-part-3/)***

_________________________

_________________________

### 🚀 [Guided project - Develop conditional branching and looping structures in C#](https://learn.microsoft.com/en-gb/training/modules/guided-project-develop-conditional-branching-looping/)

_________________________

👀 Guided Project: [My Repo](./GuidedProjectBranchingLooping/)

####  ✔️ **Learning Objectives:**


Gain experience developing a console app that implements selection and iteration statements to achieve app specifications.

* Use Visual Studio Code to develop a C# console application that uses a combination of: 
    * selection and iteration statements to: 
        * implement logical workflows in accordance with: 
            * supplied application data and user interactions.
* Evaluate the underlying conditions and make an informed decision when choosing: 
    * between `if-elseif-else` and `switch` statements, and 
    * between `foreach`, `for`, `while`, and `do` iteration statements.
* Scope variables at an appropriate level within an application.

#### **Introduction to Project:**

![Alt text](\Assets\image-3.png)

#### **Prepare**

* Use VSC to develope application
* Use Boolean expressions, selection statements, iteration statements to implement featres from design spec

##### Project Overview:

![Alt text](\Assets\image-4.png)

![Alt text](\Assets\image.png)


##### Setup:

![Alt text](\Assets\image-1.png)

#### 🚀 **[Exercise - Build sample data and menu selection loops](https://learn.microsoft.com/en-gb/training/modules/guided-project-develop-conditional-branching-looping/3-exercise-build-data-menu-loops)**

![Alt text](\Assets\image-5.png)

##### ***1. Code Review: Review the contents of the [Program.cs](..\GuidedProjectBranchingLooping\Starter\Program.cs) file in the [GuidedProjectBranchingLooping](..\GuidedProjectBranchingLooping) directory***

* The `Program.cs` file contains a preliminary version of the application that you update during this module. 
* The existing code generates sample data for the application and displays a list of menu options. 
* The menu options represent the main features of your application.

4. Notice that the top portion of your code begins with some variable declarations.

    * At the top of the file, you see a comment line and a list of variables. 
    * These variables, `animalSpecies` through `animalNickname`, are used to hold the values of the pet characteristics. 
    * Later in the code you assign the characteristic values to a multidimensional string array named ourAnimals. 
    * Each of these variables is initialized to contain a zero length string `""`. The ourAnimals array is declared a little further down in the code.
    * The next group of variables is a mix of string and int variables that you'll use to generate sample data, read user input, and establish exit criteria for your main program loop. 
    * You may have noticed the code line `string?` `readResult;.
    * When used in a variable declaration like this, the `?` character defines a nullable type variable. 
    * When reading user entered values with the `Console.ReadLine()` method, it's best to use a nullable type.
    * The final variable is a two-dimensional string array named ourAnimals. 
    * Since you're instantiating the array without initializing any values, you use the new operator (the new operator is used to create a new instance of a type). 
    * The number of rows is defined by maxPets, which has been initialized to eight. 
    * The number of characteristics that you're storing is six, the string variables that you examined above.

In [71]:
// the ourAnimals array will store the following: 
string animalSpecies = "";
string animalID = "";
string animalAge = "";
string animalPhysicalDescription = "";
string animalPersonalityDescription = "";
string animalNickname = "";

// variables that support data entry
int maxPets = 8; // generates sample data
string? readResult; // null type variable (used for Console.ReadLine())
string menuSelection = ""; // exit critera for main program loop

// array used to store runtime data, there is no persisted data
// instantiate object class instance
string[,] ourAnimals = new string[maxPets, 6]; // new used instead of initialising variable values
    // creates a 2D array
    // rows = maxPets = 8
    // columns = characteristics = 6

    // result will look like this:
    // ourAnimals[[],[],[]]

5. Scroll down to examine the for loop that contains an `if-elseif-else` selection construct.

6. Notice that the for loop uses the `maxPets` variable to establish an upper bound on the number of iterations.

7. Notice that the `if-elseif-else` construct selectively branches your code based on pet characteristics.
    <br>

    The `if-elseif-else` construct is used to define different values for the first four iterations of the for loop. After the fourth iteration, all characteristics are assigned an empty, or zero-length, string.
    <br>

    The values of the animal characteristic variables are assigned to the ourAnimals array at the bottom of the for loop.

In [72]:
// TODO: Convert the if-elseif-else construct to a switch statement

// create some initial ourAnimals array entries
for (int i = 0; i < maxPets; i++) // maxPets used as upper bounds for number of iterations
{
    // if-elseif-else selectively branches code based on pet characteristics (eg: age, nickname etc)

    if (i == 0)
    {
        animalSpecies = "dog";
        animalID = "d1";
        animalAge = "2";
        animalPhysicalDescription = "medium sized cream colored female golden retriever weighing about 65 pounds. housebroken.";
        animalPersonalityDescription = "loves to have her belly rubbed and likes to chase her tail. gives lots of kisses.";
        animalNickname = "lola";
    }
    else if (i == 1)
    {
        animalSpecies = "dog";
        animalID = "d2";
        animalAge = "9";
        animalPhysicalDescription = "large reddish-brown male golden retriever weighing about 85 pounds. housebroken.";
        animalPersonalityDescription = "loves to have his ears rubbed when he greets you at the door, or at any time! loves to lean-in and give doggy hugs.";
        animalNickname = "loki";
    }
    else if (i == 2)
    {
        animalSpecies = "cat";
        animalID = "c3";
        animalAge = "1";
        animalPhysicalDescription = "small white female weighing about 8 pounds. litter box trained.";
        animalPersonalityDescription = "friendly";
        animalNickname = "Puss";
    }
    else if (i == 3)
    {
        animalSpecies = "cat";
        animalID = "c4";
        animalAge = "?";
        animalPhysicalDescription = "";
        animalPersonalityDescription = "";
        animalNickname = "";
    }
    else
    {
        animalSpecies = "";
        animalID = "";
        animalAge = "";
        animalPhysicalDescription = "";
        animalPersonalityDescription = "";
        animalNickname = "";
    }

    // populates element in array
    ourAnimals[i, 0] = "ID #: " + animalID;
    ourAnimals[i, 1] = "Species: " + animalSpecies;
    ourAnimals[i, 2] = "Age: " + animalAge;
    ourAnimals[i, 3] = "Nickname: " + animalNickname;
    ourAnimals[i, 4] = "Physical description: " + animalPhysicalDescription;
    ourAnimals[i, 5] = "Personality: " + animalPersonalityDescription;
}

8. Scroll to the bottom of the code file, and then examine the code that's used to display the menu options and read the user selection.

9. Notice that you assign the value returned by the `Console.ReadLine()` method to the nullable string readResult.
    <br>

    * Using a nullable string is best practice for capturing input from the `ReadLine()` method. 
    * Once you verify that the input value isn't `null`, you assign the value to a standard string variable named `menuSelection`. 
    * This process enables you to evaluate the menu selection value without a concern for `null` values. 
    * Many of the methods that accept strings as an input parameter will generate an error if they're passed a null value. 
    * If you don't follow this input pattern, the code compiler is likely to generate a warning when you build your project.
    <br>

    The final lines of the `Program.cs` file echo the menu option selection and then pause execution until the `Enter` key is pressed.

In [ ]:
// display the top-level menu options

Console.Clear();

Console.WriteLine("Welcome to the Contoso PetFriends app. Your main menu options are:");
Console.WriteLine(" 1. List all of our current pet information");
Console.WriteLine(" 2. Add a new animal friend to the ourAnimals array");
Console.WriteLine(" 3. Ensure animal ages and physical descriptions are complete");
Console.WriteLine(" 4. Ensure animal nicknames and personality descriptions are complete");
Console.WriteLine(" 5. Edit an animal’s age");
Console.WriteLine(" 6. Edit an animal’s personality description");
Console.WriteLine(" 7. Display all cats with a specified characteristic");
Console.WriteLine(" 8. Display all dogs with a specified characteristic");
Console.WriteLine();
Console.WriteLine("Enter your selection number (or type Exit to exit the program)");

readResult = Console.ReadLine();
if (readResult != null)
{
    menuSelection = readResult.ToLower();
}

Console.WriteLine($"You selected menu option {menuSelection}.");
Console.WriteLine("Press the Enter key to continue");

// pause code execution
readResult = Console.ReadLine();

##### ***2. Sample Data: Convert if statement to switch statement***

In [73]:
// replace this with a switch statement:

    if (i == 0)
    {
        animalSpecies = "dog";
        animalID = "d1";
        animalAge = "2";
        animalPhysicalDescription = "medium sized cream colored female golden retriever weighing about 65 pounds. housebroken.";
        animalPersonalityDescription = "loves to have her belly rubbed and likes to chase her tail. gives lots of kisses.";
        animalNickname = "lola";
    }

Error: (3,9): error CS0103: The name 'i' does not exist in the current context

In [ ]:
    switch (i)
    {
        case 0:
            animalSpecies = "dog";
            animalID = "d1";
            animalAge = "2";
            animalPhysicalDescription = "medium sized cream colored female golden retriever weighing about 65 pounds. housebroken.";
            animalPersonalityDescription = "loves to have her belly rubbed and likes to chase her tail. gives lots of kisses.";
            animalNickname = "lola";
            break;
    }

| IF-ELSE | SWITCH |
|-|-|
| `if (i ==0)` | `case 0:` |

In [74]:
// change these if statements to switches

for (int i = 0; i < maxPets; i++)
{
    if (i == 0)
    {
        animalSpecies = "dog";
        animalID = "d1";
        animalAge = "2";
        animalPhysicalDescription = "medium sized cream colored female golden retriever weighing about 65 pounds. housebroken.";
        animalPersonalityDescription = "loves to have her belly rubbed and likes to chase her tail. gives lots of kisses.";
        animalNickname = "lola";
    }
        else if (i == 1)
    {
        animalSpecies = "dog";
        animalID = "d2";
        animalAge = "9";
        animalPhysicalDescription = "large reddish-brown male golden retriever weighing about 85 pounds. housebroken.";
        animalPersonalityDescription = "loves to have his ears rubbed when he greets you at the door, or at any time! loves to lean-in and give doggy hugs.";
        animalNickname = "loki";
    }
    else if (i == 2)
    {
        animalSpecies = "cat";
        animalID = "c3";
        animalAge = "1";
        animalPhysicalDescription = "small white female weighing about 8 pounds. litter box trained.";
        animalPersonalityDescription = "friendly";
        animalNickname = "Puss";
    }
    else if (i == 3)
    {
        animalSpecies = "cat";
        animalID = "c4";
        animalAge = "?";
        animalPhysicalDescription = "";
        animalPersonalityDescription = "";
        animalNickname = "";
    }
    else
    {
        animalSpecies = "";
        animalID = "";
        animalAge = "";
        animalPhysicalDescription = "";
        animalPersonalityDescription = "";
        animalNickname = "";
    }
    // populates element in array

    ourAnimals[i, 0] = "ID #: " + animalID;
    ourAnimals[i, 1] = "Species: " + animalSpecies;
    ourAnimals[i, 2] = "Age: " + animalAge;
    ourAnimals[i, 3] = "Nickname: " + animalNickname;
    ourAnimals[i, 4] = "Physical description: " + animalPhysicalDescription;
    ourAnimals[i, 5] = "Personality: " + animalPersonalityDescription;
}

In [75]:
// switch statements:

// create some initial ourAnimals array entries
for (int i = 0; i < maxPets; i++)
{
    switch (i)
    {
        case 0:
            animalSpecies = "dog";
            animalID = "d1";
            animalAge = "2";
            animalPhysicalDescription = "medium sized cream colored female golden retriever weighing about 65 pounds. housebroken.";
            animalPersonalityDescription = "loves to have her belly rubbed and likes to chase her tail. gives lots of kisses.";
            animalNickname = "lola";
            break;

        case 1:
            animalSpecies = "dog";
            animalID = "d2";
            animalAge = "9";
            animalPhysicalDescription = "large reddish-brown male golden retriever weighing about 85 pounds. housebroken.";
            animalPersonalityDescription = "loves to have his ears rubbed when he greets you at the door, or at any time! loves to lean-in and give doggy hugs.";
            animalNickname = "loki";
            break;
        
        case 2:
            animalSpecies = "cat";
            animalID = "c3";
            animalAge = "1";
            animalPhysicalDescription = "small white female weighing about 8 pounds. litter box trained.";
            animalPersonalityDescription = "friendly";
            animalNickname = "Puss";
            break;

        case 3:
            animalSpecies = "cat";
            animalID = "c4";
            animalAge = "?";
            animalPhysicalDescription = "";
            animalPersonalityDescription = "";
            animalNickname = "";
            break;

        default:
            animalSpecies = "";
            animalID = "";
            animalAge = "";
            animalPhysicalDescription = "";
            animalPersonalityDescription = "";
            animalNickname = "";
            break;
    }

    // populates element in array

    ourAnimals[i, 0] = "ID #: " + animalID;
    ourAnimals[i, 1] = "Species: " + animalSpecies;
    ourAnimals[i, 2] = "Age: " + animalAge;
    ourAnimals[i, 3] = "Nickname: " + animalNickname;
    ourAnimals[i, 4] = "Physical description: " + animalPhysicalDescription;
    ourAnimals[i, 5] = "Personality: " + animalPersonalityDescription;
}

* execution of a switch section isn't permitted to "fall-through". `break` is used to prevent this.

![Alt text](\Assets\image-6.png)

* right click on: `CsharpProjects\GuidedProjectBranchingLooping\Starter`
* `Open in Integrated Terminal`
* `dotnet build`
* `dotnet run`
<br>
![Alt text](\Assets\image-8.png)

##### ***3. Menu Loop: Create program menu loop***

* build a `do` loop that surrounds the menu options and the code that reads user input
* why? ensures main menu refreshes for each user, each time they make a selection
* iterates until user exits program

![Alt text](\Assets\image-9.png)

In [ ]:
do
{
    // display the top-level menu options

    Console.Clear();

    Console.WriteLine("Welcome to the Contoso PetFriends app. Your main menu options are:");
    Console.WriteLine(" 1. List all of our current pet information");
    Console.WriteLine(" 2. Add a new animal friend to the ourAnimals array");
    Console.WriteLine(" 3. Ensure animal ages and physical descriptions are complete");
    Console.WriteLine(" 4. Ensure animal nicknames and personality descriptions are complete");
    Console.WriteLine(" 5. Edit an animal’s age");
    Console.WriteLine(" 6. Edit an animal’s personality description");
    Console.WriteLine(" 7. Display all cats with a specified characteristic");
    Console.WriteLine(" 8. Display all dogs with a specified characteristic");
    Console.WriteLine();
    Console.WriteLine("Enter your selection number (or type Exit to exit the program)");

    readResult = Console.ReadLine();
    if (readResult != null)
    {
        menuSelection = readResult.ToLower();
    }

    Console.WriteLine($"You selected menu option {menuSelection}.");
    Console.WriteLine("Press the Enter key to continue");

    // pause code execution
    readResult = Console.ReadLine();
    
} while (menuSelection != "exit");

* fix formatting of code in VSC: `windows` + `f` || right click + `format document`

##### ***4. Menu Selections: Write switch statement for menu selections***

* add a `switch` statement for menu selection

![Alt text](\Assets\image-7.png)

In [ ]:
    // Console.WriteLine($"You selected menu option {menuSelection}.");
    // Console.WriteLine("Press the Enter key to continue");

    // // pause code execution
    // readResult = Console.ReadLine();

    switch(menuSelection)
    {
        case "1": // checking for a string value (as above in do)
            break;

        case "2":
            break;

        case "3":
            break;

        case "4":
            break;

        case "5":
            break;

        case "6":
            break;

        case "7":
            break;

        case "8":
            break;

        default: // optional
            break; 
    }
    
} while (menuSelection != "exit");

##### ***5. Code Branches: Write placeholder code for each case of the switch statement***

* update execution paths created by switch statement --> which will acknowledge user selection
* feedback is important for user interraction

    1. Add feedback block for each case + meaningful comments
    2. Edit case 3 + 4 with custom feedback messages
    3. don't add feedback message to default

![Alt text](\Assets\image-10.png)


![Alt text](\Assets\image-11.png)

In [ ]:
    switch(menuSelection)
    {
        case "1":  // "1" = checking for a string value (as above in do)
            // List all of our current pet information
            Console.WriteLine("this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();
            break;

        case "2":
            // Add a new animal friend to the ourAnimals array
            Console.WriteLine("this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

        case "3":
            // Ensure animal ages and physical descriptions are complete
            Console.WriteLine("Challenge Project: this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

        case "4":
            // Ensure animal nicknames and personality descriptions are complete
            Console.WriteLine("UNDER CONSTRUCTION: this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

        case "5":
            // Edit an animal's age
            Console.WriteLine("this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

        case "6":
            // Edit an animal's personality description
            Console.WriteLine("this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

        case "7":
            // Display all cats with a specified characteristic
            Console.WriteLine("this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

        case "8":
            // Display all dogs with a specified characteristic
            Console.WriteLine("this app feature is coming soon - please check back to see progress.");
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

        default: // optional
            break; 
    }
    
} while (menuSelection != "exit");


##### ***6. Verification Test: Check your work***

![Alt text](\Assets\image-12.png)

![Alt text](\Assets\image-13.png)

#### 🚀 **[Exercise - Write code to display all ourAnimals array data](https://learn.microsoft.com/en-gb/training/modules/guided-project-develop-conditional-branching-looping/4-exercise-display-all-array-data)**

![Alt text](\Assets\image-14.png)

##### ***1. Outer Loop: Build a loop to iterate through the ourAnimals array***

* create the outer `for` loop --> to ierate through ourAnimals array
* consider difference between `for` and `foreach`

    1. delete the `WriteLine` statements to display coming soon message in `menuSelection` switch statements and leave blank code line
        * this isn't required as the action will replace the temp feedback statement
    2. create


    ![Alt text](\Assets\image-15.png)

In [ ]:
    switch(menuSelection)
    {
        case "1":  // "1" = checking for a string value (as above in do)
            // List all of our current pet information
            // Console.WriteLine("this app feature is coming soon - please check back to see progress.");
            for (int i=0; 1 < maxPets; i++)
            {

            }
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();
            break;

Alternative to for loop:

![Alt text](./Assets/image-16.png)

##### ***2. Data Check: Check for existing pet data and display result***

In [ ]:
// check pets are assigned pet data - use if statement

for (int i=0; i < maxPets; i++)
{
    if (ourAnimals[i,0] != "ID #: ") // if first row of array item doesn't have ID
    {
        Console.WriteLine(ourAnimals[i, 0]);
    }
}

##### ***3. Inner Loop: Display all pet characteristics for pets with data assigned***

In [ ]:
for (int i = 0; i < maxPets; i++)
{
    if (ourAnimals[i, 0] != "ID #: ") // animals
    {
        Console.WriteLine();
        for (int j = 0; j < 6; j++) // characteristics
        {
            Console.WriteLine(ourAnimals[i, j]);
        }
    }
}

##### ***4. Verification Test: Check your work***

![Alt text](\Assets\image-17.png)

#### 🚀 **[Exercise - Build and test a loop for entering new pet data](https://learn.microsoft.com/en-gb/training/modules/guided-project-develop-conditional-branching-looping/5-exercise-build-loop-new-array-data)**

![Alt text](\Assets\image-18.png)

##### ***1. Calculate petCount: Count the number of pets in the ourAnimals array***

Establish the exit criteria for your data entry loop and you create a for loop that can be used to count the number of pets in ourAnimals that have assigned data.

1. remove "coming soon" writeline in case 2 in switch(menuSelection)
2. declare variables for calculating number of pets

![Alt text](./Assets/image2.png)

In [ ]:
        case "2":
            // Add a new animal friend to the ourAnimals array
            
            // declare variables for  calculation
            string anotherPet = "y";
            int petCount = 0;

            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

* `anotherPet` - initialised w/ y or n -> used in while loop
* `petCount` - number of animals with assigned pet characteristcs. Assigned a calculated value outside while loop and then incremented each time new animal enters ourAnimals array

Scoping variables

![Alt text](./Assets/image-scoping.png)

1. Create a loop that iterates through the animals in the `ourAnimals` array
2. inside the for loop, enter an if statement to check `ourAnimals` if pet characteristic data has been assigned to an animal
3. add an increment by 1 in `petCount`


![Alt text](./Assets/image-20.png)

In [ ]:
        case "2":
            // Add a new animal friend to the ourAnimals array
            // declare variables for  calculation
            string anotherPet = "y";
            int petCount = 0;

            // for each animal in ourAnimals, add +1 to petCount
            for (int i = 0; i < maxPets; i++)
            {
                if (ourAnimals[i, 0] != "ID #: ")
                {
                    petCount += 1;
                }
            }
            
            Console.WriteLine("Press the Enter key to continue.");
            readResult = Console.ReadLine();        
            break;

##### ***2. Conditional messages: ?***

##### ***3. Outer loop: ?***

##### ***4. Exit criteria: ?***

##### ***5. Verification test: ?***

_________________________

## ***[Work with variable data in C# console applications<br> (Get started with C#, Part 4)](https://learn.microsoft.com/en-us/training/paths/get-started-c-sharp-part-4/)***

_________________________

_________________________

## ***[Create methods in C# console applications<br> (Get started with C#, Part 5)](https://learn.microsoft.com/en-us/training/paths/get-started-c-sharp-part-5/)***

_________________________

_________________________

## ***[Debug C# console applications<br> (Get started with C#, Part 6)](https://learn.microsoft.com/en-us/training/paths/get-started-c-sharp-part-6/)***

_________________________

_________________________

### [Store and retrieve data using literal and variable values in C#](https://learn.microsoft.com/en-us/training/modules/csharp-literals-variables/)

_________________________

##### Random() & Next()

1. `Random dice = new Random();` --> create an object/instance of the `Random` class --> name it `dice` as a variable
2. `dice.Next(start,length)` --> call the dice variable and use the `Next` method

In [2]:
Random dice = new Random();  // new: create object/instance of Random class as dice variable
int roll = dice.Next(1,7); 
Console.WriteLine(roll);  // stateless method within Console. class

4
